### DISCUSSION TITLE
How can I do a long running task in Vercel or break it up into many tasks?

### DISCUSSION THREAD

**Discussion by offchan42** [vercel/community#3412] [Help] [answered]
## Problem

I'm trying to build an app that generates AI images.
The process is like this
1. user write a text prompt and submit it using POST method
2. the server calls an external API to an AI endpoint that generates the image. This is slow.
3. the server awaits for the generated image and sends it to another AI endpoint that will upscale the image. This is also slow
4. the server awaits for the upscaled image and returns it to the user as a HTTP response.

All this happens in one request from the user. When I do this locally, it's fine. But when I do it using Vercel serverless function, I got timeout issue so it prompted me to think about how to do it better.

## Potential solutions and questions

1. I'm thinking of using edge functions to stream data to user to prevent timeout but it sounds bad because my data is not like a streaming ChatGPT text. It's just 2 steps of image processing. And I hate encoding/decoding data between text and binary when using `ReadableStream`.
2. Ideally I want to run the 4 steps above without timeout and maybe even without needing to send anything back to the user. I think that I could just push the generated image into a database (something like Supabase) and then let the client subscribes to the database event to get generated images in real-time.
    1. How do I achieve this long-running function? 
    2. Is database subscription a good approach or is there another way for the client to subscribe to the data?

BTW, the AI endpoint also have a webhook support. It can calls an API after it finishes generating images. I'm not sure how to elegantly use this functionality with Vercel though. 
I'm a newbie to web dev in general, if you have a better system design pattern for this 4-step process, please let me know. Is Vercel even the right solution? Thank you.

**Comment by offchan42** [comment, marked as answer]
I just realized that I could have asked GPT-4 first. So this is answer from GPT-4 (if you have slightly different opinion, please let me know):

Based on your description, you're right that the long-running process in a single request doesn't fit well with the serverless model of Vercel. Here's an approach that could potentially solve your problem:

1. **User submits a text prompt:** When the user submits the text prompt, you create a record in your database (like Supabase) with a status of 'pending' or similar. You return the ID of this record to the user. The user's client can then poll this record for updates.

2. **Call to the AI endpoint:** Your serverless function then makes a request to the AI endpoint to generate the image. You pass the ID of the database record as part of the request. This function then ends, avoiding any timeout issues.

3. **Webhook from the AI endpoint:** The AI endpoint calls a separate serverless function on your platform when the image is ready. This function retrieves the ID of the database record from the request, and updates the record with the generated image and changes the status to 'processing'.

4. **Upscale the image:** You then make a request to the second AI endpoint to upscale the image. You again pass the ID of the database record as part of the request. This function then ends, again avoiding any timeout issues.

5. **Webhook from the AI upscaling endpoint:** The AI upscaling endpoint calls another serverless function when the upscaling is complete. This function retrieves the ID of the database record from the request, and updates the record with the upscaled image and changes the status to 'complete'.

6. **User retrieves the image:** The user's client, which has been polling the database record, sees that the status has changed to 'complete', and retrieves the upscaled image.

This approach has several advantages:

- It avoids any timeout issues, as no single function is waiting for a long-running process to complete.
- It provides feedback to the user about the status of their request.
- It makes use of the webhook functionality of the AI endpoints, which is designed for exactly this kind of scenario.

As for your question about whether Vercel is the right solution, it should be able to handle this kind of workflow. However, if you find that you're frequently running into its limitations, you might want to consider a more flexible platform like AWS, Google Cloud, or Azure. These platforms have more complex pricing models, but they also offer more control and higher limits.

**Comment by anirudhbg** [comment]
I have been using https://www.calen-ai.com/ - they are basically the vercel for long running tasks! super easy to setup

**Reply by offchan42** [reply]
I am currently using https://www.inngest.com/
